test/common/grafana_routes.go

```diff
@@ -86,56 +89,78 @@ func TestGrafanaExternalRouteDashboardExist(t *testing.T, ctx *TestingContext) {
 			APIVersion: "rbac.authorization.k8s.io/v1",
 		},
 	}
-	ctx.Client.Create(goctx.TODO(), binding)
-	defer ctx.Client.Delete(goctx.TODO(), binding)
-	ctx.Client.Get(goctx.TODO(), k8sclient.ObjectKey{Name: serviceAccountName, Namespace: grafanaNamespace}, serviceAccount)
-	secretName := serviceAccount.Secrets[0].Name
-	if secretName == serviceAccount.ImagePullSecrets[0].Name {
-		secretName = serviceAccount.Secrets[1].Name
+	err = ctx.Client.Create(goctx.TODO(), binding)
+	if err != nil {
+		t.Fatal("failed to create clusterRoleBinding", err)
 	}
-	secret := &corev1.Secret{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      secretName,
-			Namespace: grafanaNamespace,
-		},
+	defer ctx.Client.Delete(goctx.TODO(), binding)
+	err = ctx.Client.Get(goctx.TODO(), k8sclient.ObjectKey{Name: serviceAccountName, Namespace: grafanaNamespace}, serviceAccount)
+	if err != nil {
+		t.Fatal("failed to get serviceAccount", err)
 	}
-	ctx.Client.Get(goctx.TODO(), k8sclient.ObjectKey{Name: secretName, Namespace: grafanaNamespace}, secret)
-	token := string(secret.Data["token"])
 
 	grafanaRootHostname, err := getGrafanaRoute(ctx.Client)
 	if err != nil {
 		t.Fatal("failed to get grafana route", err)
 	}
-	//create new http client
-	httpClient, err := NewTestingHTTPClient(ctx.KubeConfig)
-	if err != nil {
-		t.Fatal("failed to create testing http client", err)
-	}
-	//get dashboards for grafana from the external route
-	grafanaDashboardsUrl := fmt.Sprintf("%s/api/search", grafanaRootHostname)
-	req, err := http.NewRequest("GET", grafanaDashboardsUrl, nil)
-	if err != nil {
-		t.Fatal("failed to create request for grafana", err)
-	}
-	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
-	dashboardResp, err := httpClient.Do(req)
-	if err != nil {
-		t.Fatal("failed to perform test request to grafana", err)
-	}
-	defer dashboardResp.Body.Close()
-	//there is an existing dashboard check, so confirm a valid response structure
-	if dashboardResp.StatusCode != http.StatusOK {
-		dumpResp, _ := httputil.DumpResponse(dashboardResp, true)
-		t.Logf("dumpResp: %q", dumpResp)
-		t.Fatalf("unexpected status code on success request, got=%+v", dashboardResp)
-	}
 
-	var dashboards []interface{}
-	if err := json.NewDecoder(dashboardResp.Body).Decode(&dashboards); err != nil {
-		t.Fatal("failed to decode grafana dashboards response", err)
+	// there are more secrets associated with a serviceAccount, one of them should be token that can be used to authenticate
+	// try one by one until the request is successful
```

Comment:
It seems a shame the specific secret can't be retrieved by name, name regex or labels.
@jhellar Is there no deterministic way to retrieve the right secret?

Reply:
Thanks for review, this should be fixed now.